### ISSUE TITLE
Reconnect to Denon after poweroff

### ISSUE THREAD

**Issue by jakolion** [foxriver76/ioBroker.denon#232]
**Describe the bug**
When the Denon was disconnected from the power supply and has been plugged in again, the adapter does not reconnect to the Denon successfully. I have to manually restart the adapter via the iobroker UI, then it works again.

**To Reproduce**
1. Make sure the adapter is connected to the Denon
2. Disconnect the power cable of the Denon
3. After a few minutes, plug it in
4. The adapter is not connecting to the Denon

**Expected behavior**
The adapter should detect the Denon after a specific time and reconnect to it.

**Screenshots & Logfiles**
Please tell me what files are needed and how to get them for further investigation.

**Versions:**
 - Adapter version: 1.12.1
 - JS-Controller version: 4.0.21
 - Node version: 14.19.1
 - Operating system: Raspbian OS
 - AVR model: x2000

Thanks!

**Comment by foxriver76** [comment]
Please show debug log of a non happening connection, the adapter is trying to reconnect silently if the error stays the same

**Comment by jakolion** [comment]
I found out how to enable the debug logs for the Denon adapter. I deleted all logs and unplugged the Denon. There was nothing logged by the adapter. At "Instances" it still says that it's "Connected to device or service".

When plugging in the Denon this stays the same, the connection is not rebuild. When the status of the Denon changes (for example turning on the Denon), the data is not updated. Nothing seems to be logged by the adapter in this state.

Now when requesting to update data (for example changing the volume of the main zone), the adapter reconnects successfully. It updates nearly all data and is able to change them. Nearly all data because the power status of the zones "denon.0.zoneMain.powerZone" and "denon.0.zone2.powerZone" in my case stays "false" which is not correct.

Attached the logs. The logs starts at "2022-03-23 10:07:25.619", which is the moment when sending a new volume after turning on the Denon. But the logging was already started before unplugging it.

[denon.log](https://github.com/foxriver76/ioBroker.denon/files/8331671/denon.log)

**Comment by foxriver76** [comment]
So according to your log, the connection stays open all the time, so of course no reconnect happens. Either the connection is closed or the connection times out if there are 35 seconds without a response from the AVR. So yes technically there is no new connection according to your log thus we do not poll and also it seems like the AVR itself is not responding to the polling commands so it should time out. Or log is cutted and AVR responses then, but in your log there is no connection loss so of course no reconnect.

**Comment by jakolion** [comment]
The AVR was unplugged for more than 15 minutes. Uploaded another log - it shows, that the previous one was not cutted at the beginning.

[denon2.log](https://github.com/foxriver76/ioBroker.denon/files/8331931/denon2.log)

Unplugged the AVR at 10:54
Plugged in, turned on and changed the volume via remote at 10:59
Request volume update via ioBroker at 11:01

The log shows a gap from 10:32 (last action before unplugging) to 11:01. So because there are not log entries trying to reach the Denon, I asume the polling is not working correctly.

And as mentioned before: after reconnect the power states of the zones are not updated.

**Comment by foxriver76** [comment]
So really nothing between 

```
2022-03-23 10:32:48.328  - [34mdebug[39m: denon.0 (16148) [INFO] ==> Message sent: BDSTATUS?
2022-03-23 11:01:06.404  - [34mdebug[39m: denon.0 (16148) [COMMAND] State Change - ID: zoneMain.volume; State: 37.5
```

Have to look how this can be

**Comment by foxriver76** [comment]
Can you confirm that there is no logging in between, because in the prev. post you said unplugged at 10:54 but no data is sent since 10:32?

**Comment by jakolion** [comment]
Yes, that's right. The AVR was turned off round about 10:32 (so it was in standby) and at 10:54 I unplugged the power cable.

**Comment by foxriver76** [comment]
Do you have Network Control configured to always?

**Comment by jakolion** [comment]
Yes, it is set to always on.

**Comment by foxriver76** [comment]
please try if 1.13.2 makes a difference

**Comment by jakolion** [comment]
I would do so, but have no experience in updating modules manually, which are not available via ioBroker yet. Can you guide me to an instruction how to do that?

**Comment by foxriver76** [comment]
`iob upgrade denon@1.13.2`

**Comment by jakolion** [comment]
It worked once. I can see the polling and this is the part where the first disconnect and reconnect was done successfully.

`2022-03-23 12:45:09.239  - [33mwarn[39m: denon.0 (24775) AVR timed out due to no response
2022-03-23 12:45:39.242  - [32minfo[39m: denon.0 (24775) [CONNECT] Trying to connect to 10.1.1.16:23
2022-03-23 12:45:42.399  - [33mwarn[39m: denon.0 (24775) AVR unreachable, check the Network Config of your AVR
2022-03-23 12:46:12.402  - [32minfo[39m: denon.0 (24775) [CONNECT] Trying to connect to 10.1.1.16:23
2022-03-23 12:46:15.514  - [34mdebug[39m: denon.0 (24775) AVR unreachable, check the Network Config of your AVR
2022-03-23 12:46:45.515  - [34mdebug[39m: denon.0 (24775) [CONNECT] Trying to connect to 10.1.1.16:23
2022-03-23 12:46:48.633  - [34mdebug[39m: denon.0 (24775) AVR unreachable, check the Network Config of your AVR
2022-03-23 12:47:18.635  - [34mdebug[39m: denon.0 (24775) [CONNECT] Trying to connect to 10.1.1.16:23
2022-03-23 12:47:18.638  - [32minfo[39m: denon.0 (24775) [CONNECT] Adapter connected to DENON-AVR: 10.1.1.16:23
2022-03-23 12:47:18.639  - [34mdebug[39m: denon.0 (24775) [CONNECT] Connected --> updating states on start`

The second time it stopped polling but the connected status did not change. It looks different from the first time. The first "Connection closed: Error: read ECONNRESET" appeared when I already reconnected the AVR.

`2022-03-23 12:53:32.926  - [34mdebug[39m: denon.0 (24775) [INFO] ==> Message sent: BDSTATUS?
[other messages]
2022-03-23 12:57:35.594  - [33mwarn[39m: denon.0 (24775) Connection closed: Error: read ECONNRESET
2022-03-23 12:58:05.595  - [32minfo[39m: denon.0 (24775) [CONNECT] Trying to connect to 10.1.1.16:23
2022-03-23 12:58:05.597  - [32minfo[39m: denon.0 (24775) [CONNECT] Adapter connected to DENON-AVR: 10.1.1.16:23
2022-03-23 12:58:05.598  - [34mdebug[39m: denon.0 (24775) [CONNECT] Connected --> updating states on start`

When I plugged in the AVR it immediately reconnects.

**Comment by jakolion** [comment]
> So really nothing between
> 
> ```
> 2022-03-23 10:32:48.328  - �[34mdebug�[39m: denon.0 (16148) [INFO] ==> Message sent: BDSTATUS?
> 2022-03-23 11:01:06.404  - �[34mdebug�[39m: denon.0 (16148) [COMMAND] State Change - ID: zoneMain.volume; State: 37.5
> ```
> 
> Have to look how this can be

**Comment by jakolion** [comment]
Didn't want to close it ...

**Comment by foxriver76** [comment]
Okay, some versions ago I removed the request interval, because I thought it would already have a slight delay due to an internal change but commands in your log got sent around the same millisecond. New version brought back the interval, try to increase the default interval in advanced settings of the adapter, especially if you are connected via wireless network or the AVR isn't the newest, I could imagine that it puts the tcp socket in a strange state.

**Comment by jakolion** [comment]
The AVR is a few years old but connected via network cable to the router.

I increased the message delay from 100ms to 250ms. The behavior is the same: the first time a timeout occurs, the second time it simply logs nothing while disconnected.

First disconnect detected after the 35 seconds you mentioned:
```
2022-03-24 13:38:28.353  - [34mdebug[39m: denon.0 (26209) [INFO] ==> Message sent: BDSTATUS?
2022-03-24 13:39:03.354  - [33mwarn[39m: denon.0 (26209) AVR timed out due to no response
```

Second disconnect detected after cable was plugged in again after round about 4 mins:
```
2022-03-24 13:40:40.385  - [34mdebug[39m: denon.0 (26209) [INFO] ==> Message sent: SPPR ?
2022-03-24 13:44:12.394  - [33mwarn[39m: denon.0 (26209) Connection closed: Error: read ECONNRESET
```

**Comment by foxriver76** [comment]
Okay, then somehow the timeout is only applied once, I will push some changes later on, then we can test

**Comment by foxriver76** [comment]
Please try

`iob url https://github.com/foxriver76/ioBroker.denon && iob u denon && iob r denon`

**Comment by jakolion** [comment]
The current version looks fine - the polling and all disconnects and reconnects are working fine. Also the power status change while disconnected was correctly updated when it reconnected successfully (the AVR was turned on while disconnected).

[denon5.log](https://github.com/foxriver76/ioBroker.denon/files/8350525/denon5.log)

Which additional commands do I have to execute on the cli when the fix is published with a new version? Because I would like to go with that version now, but I want to get updates for the stable versions of the adapter like before the testing.

Thank you very much for the quick responses and the fix!

**Comment by foxriver76** [comment]
We have nothing changed regarding your active repository, these commands were only one time installations directly from npm/github. I will publish the version later on, thanks for provding quick feedback

You can check the repository via `iob repo show`, it should mention `stable`, or simply check in the settings of the admin gui.
